File: source/clibCore/CPDocClass.vb

Imports Contensive.BaseClasses
Imports System.Runtime.InteropServices

Namespace Contensive.Core
    '
    ' comVisible to be activeScript compatible
    '
    <ComVisible(True)> _
    <ComClass(CPDocClass.ClassId, CPDocClass.InterfaceId, CPDocClass.EventsId)> _
    Public Class CPDocClass
        Inherits BaseClasses.CPDocBaseClass
        Implements IDisposable
#Region "COM GUIDs"
        Public Const ClassId As String = "414BD6A9-195F-4E0F-AE24-B7BF56749CDD"
        Public Const InterfaceId As String = "347D06BC-4D68-4DBE-82FE-B72115E24A56"
        Public Const EventsId As String = "95E8786B-E778-4617-96BA-B45C53E4AFD1"
#End Region
        '
        Private cp As CPClass
        Private cpCore As Contensive.Core.cpCoreClass
        Private LocalVars As New Collection
        Private LocalVarNameList As String
        Protected disposed As Boolean = False
        '
        Friend Sub New(ByVal cpParent As CPClass)
            MyBase.New()
            cp = cpParent
            cpCore = cp.core
        End Sub
        '
        ' dispose
        '
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                Call appendDebugLog(".dispose, dereference main, csv")
                If disposing Then
                    '
                    ' call .dispose for managed objects
                    '
                    cpCore = Nothing
                    cp = Nothing
                End If
                '
                ' Add code here to release the unmanaged resource.
                '
            End If
            Me.disposed = True
        End Sub
        '
        '
        '
        Public Overrides Property content() As String
            Get
                If True Then
                    Return cpCore.main_PageContent
                Else
                    Return ""
                End If
            End Get
            Set(ByVal value As String)
                If True Then
                    cpCore.main_PageContent = value
                End If
            End Set
        End Property

        Public Overrides ReadOnly Property navigationStructure() As String
            Get
                If True Then
                    Return cpCore.main_RenderedNavigationStructure
                Else
                    Return ""
                End If
            End Get
        End Property

        Public Overrides Property noFollow() As Boolean
            Get
                If True Then
                    Return cpCore.main_MetaContentNoFollow
                Else
                    Return False
                End If
            End Get
            Set(ByVal value As Boolean)
                If True Then
                    cpCore.main_MetaContentNoFollow = value
                End If
            End Set
        End Property

        Public Overrides ReadOnly Property pageId() As Integer
            Get
                If True Then
                    Return cpCore.main_RenderedPageID
                Else
                    Return 0
                End If
            End Get
        End Property

        Public Overrides ReadOnly Property pageName() As String
            Get
                Dim s As String
                If True Then
                    s = cpCore.main_RenderedPageName
                    If (s Is Nothing) Then
                        s = ""
                    End If
                    Return s
                Else
                    Return ""
                End If
            End Get
        End Property

        Public Overrides ReadOnly Property refreshQueryString() As String
            Get
                Dim s As String
                If True Then
                    s = cpCore.web_RefreshQueryString
                    If (s Is Nothing) Then
                        s = ""
                    End If
                    Return s
                Else
                    Return ""
                End If
            End Get
        End Property

        Public Overrides ReadOnly Property sectionId() As Integer
            Get
                If True Then
                    Return cpCore.main_RenderedSectionID
                Else
                    Return 0
                End If
            End Get
        End Property

        Public Overrides ReadOnly Property startTime() As Date
            Get
                If True Then
                    Return cpCore.main_PageStartTime
                Else
                    Return New Date()
                End If
            End Get
        End Property

        Public Overrides ReadOnly Property templateId() As Integer
            Get
                If True Then
                    Return cpCore.main_RenderedTemplateID
                Else
                    Return 0
                End If
            End Get
        End Property

        Public Overrides ReadOnly Property type() As String
            Get
                If True Then
                    Return cpCore.main_docType
                Else
                    Return ""
                End If
            End Get
        End Property
        '
        '
        '
        Public Overrides Sub addHeadJavascript(ByVal NewCode As String)
            If True Then
                Call cpCore.main_AddHeadJavascript(NewCode)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addHeadTag(ByVal HeadTag As String)
            If True Then
                Call cpCore.main_AddHeadTag(HeadTag)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addMetaDescription(ByVal MetaDescription As String)
            If True Then
                Call cpCore.main_addMetaDescription(MetaDescription)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addMetaKeywordList(ByVal MetaKeywordList As String)
            If True Then
                Call cpCore.main_addMetaKeywordList(MetaKeywordList)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addOnLoadJavascript(ByVal NewCode As String)
            If True Then
                Call cpCore.main_AddOnLoadJavascript(NewCode)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addTitle(ByVal PageTitle As String)
            If True Then
                Call cpCore.main_AddPagetitle(PageTitle)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addRefreshQueryString(ByVal Name As String, ByVal Value As String)
            If True Then
                Call cpCore.web_addRefreshQueryString(Name, Value)
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addHeadStyle(ByVal StyleSheet As String)
            If True Then
                Call addHeadTag(vbCrLf & vbTab & "<style type=""text/css"">" & vbCrLf & vbTab & vbTab & StyleSheet & vbCrLf & vbTab & "</style>")
            End If
        End Sub
        '
        '
        '
        Public Overrides Sub addHeadStyleLink(ByVal StyleSheetLink As String)
            If True Then
                Call cpCore.main_AddStylesheetLink(StyleSheetLink)
            End If
        End Sub

        Public Overrides Sub addBodyEnd(ByVal NewCode As String)
            If True Then
                cpCore.main_ClosePageHTML = cpCore.main_ClosePageHTML & NewCode
            End If
        End Sub

        Public Overrides Property body() As String
            Get
                If True Then
                    body = cpCore.main_FilterInput
                Else
                    Return ""
                End If
            End Get
            Set(ByVal value As String)
                If True Then
                    cpCore.main_FilterInput = value
                End If
            End Set
        End Property
        '
        '
        '
        Public Function getLegacyOptionStringFromVar() As String
            Dim Value As String
            Dim lcName As String
            Dim lcNames() As String
            Dim Ptr As Integer
            '
            getLegacyOptionStringFromVar = ""
            Call tp("getLegacyOptionStringFromVar enter, LocalVarNameList=" & LocalVarNameList)
            If LocalVarNameList <> "" Then
                lcNames = Split(LocalVarNameList, vbCrLf)
                If lcNames.Length > 0 Then
                    For Ptr = 0 To lcNames.Length - 1
                        lcName = lcNames(Ptr)
                        If lcName <> "" Then
                            Value = var(lcName)
                            getLegacyOptionStringFromVar = getLegacyOptionStringFromVar & "&" & encodeLegacyAddonOptionArgument(lcName) & "=" & encodeLegacyAddonOptionArgument(Value)
                        End If
                    Next
                End If
            End If
            Call tp("getLegacyOptionStringFromVar exit, getLegacyOptionStringFromVar=" & getLegacyOptionStringFromVar)
        End Function
        '
        '
        '
        Public Overrides ReadOnly Property siteStylesheet() As String
            Get
                Return cpCore.csv_getStyleSheet()
            End Get
        End Property
        '
        '------------------------------------------------------------------------------------------------------------
        '   Encodes an argument in an Addon OptionString (QueryString) for all non-allowed characters
        '       call this before parsing them together
        '       call decodeAddonOptionArgument after parsing them apart
        '
        '       Arg0,Arg1,Arg2,Arg3,Name=Value&Name=VAlue[Option1|Option2]
        '
        '       This routine is needed for all Arg, Name, Value, Option values
        '
        '------------------------------------------------------------------------------------------------------------
        '
        Friend Function encodeLegacyAddonOptionArgument(ByVal Arg As String) As String
            Dim a As String
            encodeLegacyAddonOptionArgument = ""
            If Arg <> "" Then
                a = Arg
                a = Replace(a, vbCrLf, "#0013#")
                a = Replace(a, vbLf, "#0013#")
                a = Replace(a, vbCr, "#0013#")
                a = Replace(a, "&", "#0038#")
                a = Replace(a, "=", "#0061#")
                a = Replace(a, ",", "#0044#")
                a = Replace(a, """", "#0034#")
                a = Replace(a, "'", "#0039#")
                a = Replace(a, "|", "#0124#")
                a = Replace(a, "[", "#0091#")
                a = Replace(a, "]", "#0093#")
                a = Replace(a, ":", "#0058#")
                encodeLegacyAddonOptionArgument = a
            End If
        End Function
        '
        '------------------------------------------------------------------------------------------------------------
        '   Decodes an argument parsed from an AddonOptionString for all non-allowed characters
        '       AddonOptionString is a & delimited string of name=value[selector]descriptor
        '
        '       to get a value from an AddonOptionString, first use getargument() to get the correct value[selector]descriptor
        '       then remove everything to the right of any '['
        '
        '       call encodeaddonoptionargument before parsing them together
        '       call decodeAddonOptionArgument after parsing them apart
        '
        '       Arg0,Arg1,Arg2,Arg3,Name=Value&Name=VAlue[Option1|Option2]
        '
        '       This routine is needed for all Arg, Name, Value, Option values
        '
        '------------------------------------------------------------------------------------------------------------
        '
        Friend Function decodeLegacyAddonOptionArgument(ByVal EncodedArg As String) As String
            Dim a As String
            '
            decodeLegacyAddonOptionArgument = ""
            If EncodedArg <> "" Then
                a = EncodedArg
                a = Replace(a, "#0058#", ":")
                a = Replace(a, "#0093#", "]")
                a = Replace(a, "#0091#", "[")
                a = Replace(a, "#0124#", "|")
                a = Replace(a, "#0039#", "'")
                a = Replace(a, "#0034#", """")
                a = Replace(a, "#0044#", ",")
                a = Replace(a, "#0061#", "=")
                a = Replace(a, "#0038#", "&")
                a = Replace(a, "#0013#", vbCrLf)
                decodeLegacyAddonOptionArgument = a
            End If
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function GetProperty(PropertyName As String, Optional DefaultValue As String = "") As String
            Dim s As String = ""
            Try
                '
                ' refactor to do the work directly
                '
                If isVar(PropertyName) Then
                    s = var(PropertyName)
                Else
                    s = DefaultValue
                End If
            Catch ex As Exception

            End Try
            Return s
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function GetBoolean(PropertyName As String, Optional DefaultValue As String = "") As Boolean
            Return cp.Utils.EncodeBoolean(GetProperty(PropertyName, DefaultValue))
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function GetDate(PropertyName As String, Optional DefaultValue As String = "") As Date
            Return cp.Utils.EncodeDate(GetProperty(PropertyName, DefaultValue))
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function GetInteger(PropertyName As String, Optional DefaultValue As String = "") As Integer
            Return cp.Utils.EncodeInteger(GetProperty(PropertyName, DefaultValue))
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function GetNumber(PropertyName As String, Optional DefaultValue As String = "") As Double
            Return cp.Utils.EncodeNumber(GetProperty(PropertyName, DefaultValue))
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function GetText(FieldName As String, Optional DefaultValue As String = "") As String
            Return GetProperty(FieldName, DefaultValue)
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Function IsProperty(FieldName As String) As Boolean
            Dim s As Boolean
            '
            '
            ' refactor to do the work directly
            '
            s = False
            Try
                s = isVar(FieldName)
            Catch ex As Exception

            End Try
            Return s
        End Function
        '
        '=======================================================================================================
        '
        '=======================================================================================================
        '
        Public Overrides Sub SetProperty(FieldName As String, FieldValue As String)
            '
            ' refactor to do the work directly
            '
            Try
                var(FieldName) = FieldValue
            Catch ex As Exception

            End Try
        End Sub
        '
        '=======================================================================================================
        ' Deprecated ------------ us GetProperty and Setproperty
        ' IsVar
        '   like GlobalVar, but it includes the OptionString
        '   a set Var adds an etry to LocalVars
        '=======================================================================================================
        '
        Public Overrides Property var(ByVal Name As String) As String
            Get
                Dim lcName As String
                If Name = "" Then
                    var = ""
                Else
                    lcName = Name.ToLower
                    If LocalVars.Contains(lcName) Then
                        var = LocalVars(lcName)
                    Else
                        var = globalVar(lcName)
                    End If
                End If
                Call tp("Property var get exit, " & Name & "=" & var)
            End Get
            Set(ByVal value As String)
                Dim lcName As String
                'Dim valueObj As String = ""
                'If Not value Is Nothing Then
                '    valueObj = value
                'End If
                value = EncodeText(value)
                Call appendDebugLog("var set, " & Name & "=" & value.ToString)
                If Name <> "" Then
                    lcName = Name.ToLower
                    If LocalVars.Contains(lcName) Then
                        Call LocalVars.Remove(lcName)
                        If value = "" Then
                            LocalVarNameList = Replace(LocalVarNameList, vbCrLf & lcName, "", , , CompareMethod.Text)
                        Else
                            Call LocalVars.Add(value, lcName)
                            Call tp("Property var set, name found in localVars, removed and re-added, LocalVarNameList=" & LocalVarNameList)
                        End If
                    ElseIf value <> "" Then
                        Call LocalVars.Add(value, lcName)
                        LocalVarNameList = LocalVarNameList & vbCrLf & lcName
                        Call tp("Property var set, name not found in localVars so it was added, LocalVarNameList=" & LocalVarNameList)
                    End If
                End If
            End Set
        End Property
        '
        '=======================================================================================================
        ' Deprecated ------------ us GetProperty and Setproperty
        ' GlobalVar
        '   Like ViewingProperties but includes stream
        '   returns
        '       matches to ViewingProperties
        '       then matches to Stream
        '=======================================================================================================
        '
        Public Overrides Property globalVar(ByVal Name As String) As String
            Get
                Dim lcName As String
                If True Then
                    If Name = "" Then
                        globalVar = ""
                    Else
                        lcName = Name.ToLower
                        If cpCore.main_IsViewingProperty(lcName) Then
                            globalVar = cpCore.main_GetViewingProperty(lcName)
                        Else
                            globalVar = cpCore.doc_getText(lcName)
                        End If
                    End If
                Else
                    globalVar = ""
                End If
            End Get
            Set(ByVal value As String)
                If True Then
                    Call cpCore.main_SetViewingProperty(Name.ToLower, value)
                End If
            End Set
        End Property
        '
        '=======================================================================================================
        ' Deprecated ------------ us GetProperty and Setproperty
        ' IsGlobalVar 
        '   returns true if
        '       IsViewingProperties is true or InStream is true
        '=======================================================================================================
        '
        Public Overrides ReadOnly Property isGlobalVar(ByVal Name As String) As Boolean
            Get
                Dim lcName As String
                If True Then
                    If Name = "" Then
                        isGlobalVar = ""
                    Else
                        lcName = Name.ToLower
                        isGlobalVar = cpCore.main_IsViewingProperty(lcName)
                        If Not isGlobalVar Then
                            isGlobalVar = cpCore.main_InStream(lcName)
                        End If
                    End If
                Else
                    isGlobalVar = False
                End If
            End Get
        End Property
        '
        '=======================================================================================================
        ' Deprecated ------------ us GetProperty and Setproperty
        ' IsVar
        '   returns true if
        '       matches to GlobalVars or LocalVars collections
        '=======================================================================================================
        '
        Public Overrides ReadOnly Property isVar(ByVal Name As String) As Boolean
            Get
                Dim lcName As String
                If Name = "" Then
                    isVar = False
                Else
                    lcName = Name.ToLower
                    isVar = LocalVars.Contains(lcName)
                    If Not isVar Then
                        isVar = isGlobalVar(lcName)
                    End If
                End If
            End Get
        End Property
        '
        '
        '
        Public Overrides ReadOnly Property IsAdminSite As Boolean
            Get
                Dim returnIsAdmin As Boolean = False
                Try
                    returnIsAdmin = (InStr(1, cp.Request.PathPage, cp.Site.GetText("adminUrl"), vbTextCompare) <> 0)
                Catch ex As Exception
                    Call cp.core.handleException(ex, "unexpected error in IsAdminSite")
                End Try
                Return returnIsAdmin
            End Get
        End Property
        '
        '=======================================================================================================
        '
        ' debugging -- append to logfile
        '
        '=======================================================================================================
        '
        Private Sub appendDebugLog(ByVal copy As String)
            'My.Computer.FileSystem.WriteAllText("c:\clibCpDocDebug.log", Now & " - cp.doc, " & copy & vbCrLf, True)
            ' 'My.Computer.FileSystem.WriteAllText(System.AppDocmc.main_CurrentDocmc.main_BaseDirectory() & "cpLog.txt", Now & " - " & copy & vbCrLf, True)
        End Sub
        '
        '=======================================================================================================
        '
        ' debugging -- testpoint
        '
        '=======================================================================================================
        '
        Private Sub tp(ByVal msg As String)
            'Call appendDebugLog(msg)
        End Sub
        '
        '=======================================================================================================
        '
        ' IDisposable support
        '
        '=======================================================================================================
        '
#Region " IDisposable Support "
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class
End Namespace

File: source/core/Views/CPDocClass.cs

using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using Contensive.Processor;
using Contensive.Processor.Models.DbModels;
using Contensive.Processor.Controllers;
using static Contensive.Processor.Controllers.genericController;
using static Contensive.Processor.constants;
//

namespace Contensive.Processor {
    public class CPDocClass : BaseClasses.CPDocBaseClass, IDisposable {
        #region COM GUIDs
        public const string ClassId = "414BD6A9-195F-4E0F-AE24-B7BF56749CDD";
        public const string InterfaceId = "347D06BC-4D68-4DBE-82FE-B72115E24A56";
        public const string EventsId = "95E8786B-E778-4617-96BA-B45C53E4AFD1";
        #endregion
        //
        private CPClass cp;
        private Contensive.Processor.Controllers.CoreController core;
        protected bool disposed = false;
        //
        //====================================================================================================
        /// <summary>
        /// constructor
        /// </summary>
        /// <param name="cpParent"></param>
        public CPDocClass(CPClass cpParent) : base() {
            cp = cpParent;
            core = cp.core;
        }
        //
        //====================================================================================================
        /// <summary>
        /// destructor
        /// </summary>
        /// <param name="disposing"></param>
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                appendDebugLog(".dispose, dereference main, csv");
                if (disposing) {
                    //
                    // call .dispose for managed objects
                    //
                    core = null;
                    cp = null;
                }
                //
                // Add code here to release the unmanaged resource.
                //
            }
            this.disposed = true;
        }
        //
        //====================================================================================================
        /// <summary>
        /// Returns the page content
        /// </summary>
        /// <returns></returns>
        public override string Content {
            get {
                return core.doc.bodyContent;
            }
            set {
                core.doc.bodyContent = value;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// deprecated.
        /// </summary>
        /// <returns></returns>
        [Obsolete("Use addon navigation.", true)]
        public override string NavigationStructure {
            get {
                return string.Empty ;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// Returns to the current value of NoFollow, set by addon execution
        /// </summary>
        /// <returns></returns>
        public override bool NoFollow {
            get {
                return core.webServer.response_NoFollow;
            }
            set {
                core.webServer.response_NoFollow = value;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the pageId
        /// </summary>
        /// <returns></returns>
        public override int PageId {
            get {
                if (core.doc.pageController.page == null) {
                    return 0;
                } else {
                    return core.doc.pageController.page.id;
                }
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the page name, set by the pagemenager addon
        /// </summary>
        /// <returns></returns>
        public override string PageName {
            get {
                if (core.doc.pageController.page == null) {
                    return string.Empty;
                } else {
                    return core.doc.pageController.page.name;
                }
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the current value of refreshquerystring 
        /// </summary>
        /// <returns></returns>
        public override string RefreshQueryString {
            get {
                return core.doc.refreshQueryString;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// Returns the value of sectionId
        /// </summary>
        [Obsolete("Section is no longer supported", true)]
        public override int SectionId {
            get {
                return 0;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// the time and date when this document was started 
        /// </summary>
        /// <returns></returns>
        public override DateTime StartTime {
            get {
                return core.doc.profileStartTime;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the id of the template, as set by the page manager
        /// </summary>
        /// <returns></returns>
        public override int TemplateId {
            get {
                if (core.doc != null) {
                    if (core.doc.pageController.template != null) {
                        return core.doc.pageController.template.id;
                    }
                }
                return 0;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the docType, set by the page manager settings 
        /// </summary>
        /// <returns></returns>
        public override string Type {
            get {
                return core.siteProperties.docTypeDeclaration;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// adds javascript code to the head of the document
        /// </summary>
        /// <param name="NewCode"></param>
        public override void AddHeadJavascript(string NewCode) {
            core.html.addScriptCode(NewCode, "");
        }
        //
        //====================================================================================================
        /// <summary>
        /// adds a javascript tag to the head of the document
        /// </summary>
        /// <param name="HeadTag"></param>
        public override void AddHeadTag(string HeadTag) {
            core.html.addHeadTag(HeadTag);
        }
        //
        //====================================================================================================
        //
        public override void AddMetaDescription(string MetaDescription) {
            core.html.addMetaDescription(MetaDescription);
        }
        //
        //====================================================================================================
        //
        public override void AddMetaKeywordList(string MetaKeywordList) {
            core.html.addMetaKeywordList(MetaKeywordList);
        }
        //
        //====================================================================================================
        //
        public override void AddOnLoadJavascript(string NewCode) {
            core.html.addScriptCode_onLoad(NewCode, "");
        }
        //
        //====================================================================================================
        //
        public override void AddTitle(string PageTitle) {
            core.html.addTitle(PageTitle);
        }
        //
        //====================================================================================================
        //
        public override void AddRefreshQueryString(string Name, string Value) {
            core.doc.addRefreshQueryString(Name, Value);
        }
        //
        //====================================================================================================
        //
        public override void AddHeadStyle(string StyleSheet) {
            AddHeadTag("\r\n\t<style type=\"text/css\">\r\n\t\t" + StyleSheet + "\r\n\t</style>");
        }
        //
        //====================================================================================================
        //
        public override void AddHeadStyleLink(string StyleSheetLink) {
            core.html.addStyleLink(StyleSheetLink, "");
        }
        //
        //====================================================================================================
        //
        public override void AddBodyEnd(string html) {
            core.doc.htmlForEndOfBody += html;
        }
        //
        //====================================================================================================
        //
        public override string Body {
            get {
                return core.doc.docBodyFilter;
            }
            set {
                core.doc.docBodyFilter = value;
            }
        }

        //
        //====================================================================================================
        //
        [Obsolete("Site styles are no longer supported. Include styles and javascript in addons.", true)]
        public override string SiteStylesheet {
            get {
                return "";
            }
        }
        //
        //====================================================================================================
        //   Decodes an argument parsed from an AddonOptionString for all non-allowed characters
        //       AddonOptionString is a & delimited string of name=value[selector]descriptor
        //
        //       to get a value from an AddonOptionString, first use getargument() to get the correct value[selector]descriptor
        //       then remove everything to the right of any '['
        //
        //       call encodeaddonoptionargument before parsing them together
        //       call decodeAddonOptionArgument after parsing them apart
        //
        //       Arg0,Arg1,Arg2,Arg3,Name=Value&Name=VAlue[Option1|Option2]
        //
        //       This routine is needed for all Arg, Name, Value, Option values
        //
        //------------------------------------------------------------------------------------------------------------
        //
        public string decodeLegacyOptionStringArgument(string EncodedArg) {
            string tempdecodeLegacyOptionStringArgument = null;
            string a = null;
            //
            tempdecodeLegacyOptionStringArgument = "";
            if (!string.IsNullOrEmpty(EncodedArg)) {
                a = EncodedArg;
                a = genericController.vbReplace(a, "#0058#", ":");
                a = genericController.vbReplace(a, "#0093#", "]");
                a = genericController.vbReplace(a, "#0091#", "[");
                a = genericController.vbReplace(a, "#0124#", "|");
                a = genericController.vbReplace(a, "#0039#", "'");
                a = genericController.vbReplace(a, "#0034#", "\"");
                a = genericController.vbReplace(a, "#0044#", ",");
                a = genericController.vbReplace(a, "#0061#", "=");
                a = genericController.vbReplace(a, "#0038#", "&");
                a = genericController.vbReplace(a, "#0013#", "\r\n");
                tempdecodeLegacyOptionStringArgument = a;
            }
            return tempdecodeLegacyOptionStringArgument;
        }
        //
        //=======================================================================================================
        //
        public override string GetProperty(string PropertyName, string DefaultValue = "") {
            if (core.docProperties.containsKey(PropertyName)) {
                return core.docProperties.getText(PropertyName);
            } else {
                return DefaultValue;
            }
        }
        //
        //=======================================================================================================
        //
        public override bool GetBoolean(string PropertyName, string DefaultValue = "") {
            return genericController.encodeBoolean(GetProperty(PropertyName, DefaultValue));
        }
        //
        //=======================================================================================================
        //
        public override DateTime GetDate(string PropertyName, string DefaultValue = "") {
            return genericController.encodeDate(GetProperty(PropertyName, DefaultValue));
        }
        //
        //=======================================================================================================
        //
        public override int GetInteger(string PropertyName, string DefaultValue = "") {
            return cp.Utils.EncodeInteger(GetProperty(PropertyName, DefaultValue));
        }
        //
        //=======================================================================================================
        //
        public override double GetNumber(string PropertyName, string DefaultValue = "") {
            return cp.Utils.EncodeNumber(GetProperty(PropertyName, DefaultValue));
        }
        //
        //=======================================================================================================
        //
        public override string GetText(string FieldName, string DefaultValue = "") {
            return GetProperty(FieldName, DefaultValue);
        }
        //
        //=======================================================================================================
        //
        public override bool IsProperty(string FieldName) {
            return core.docProperties.containsKey(FieldName);
        }
        //
        //=======================================================================================================
        //
        public override void SetProperty(string FieldName, string FieldValue) {
            core.docProperties.setProperty(FieldName, FieldValue);
        }
        // 20171229 - converted to getter and setter
        //=======================================================================================================
        //
        //public override void Var(string Index, string Value) {
        //    SetProperty( Index, Value);
        //}
        //
        //=======================================================================================================
        //
        public override string get_GlobalVar(string Index) {
            return get_Var( Index );
        }
        //
        //=======================================================================================================
        //
        public override bool get_IsGlobalVar(string Index) {
            return get_IsVar( Index );
        }
        //
        //=======================================================================================================
        //
        public override bool get_IsVar(string Index) {
            return core.docProperties.containsKey(Index);
        }
        //
        //=======================================================================================================
        //
        public override string get_Var(string Index) {
            return core.docProperties.getText(Index);
        }
        //
        //=======================================================================================================
        //
        public override void set_Var(string Index, string Value) {
            core.docProperties.setProperty(Index, Value);
        }
        //
        //=======================================================================================================
        //
        public override void set_GlobalVar(string Index, string Value) {
            //
        }
        //
        //=======================================================================================================
        //
        public override bool IsAdminSite {
            get {
                bool returnIsAdmin = false;
                try {
                    returnIsAdmin = (cp.Request.PathPage.IndexOf(cp.Site.GetText("adminUrl"), System.StringComparison.OrdinalIgnoreCase)  != -1);
                } catch (Exception ex) {
                    logController.handleError( core,ex); 
                    throw;
                }
                return returnIsAdmin;
            }
        }
        //
        //=======================================================================================================
        //
        private void appendDebugLog(string copy) {
        }
        //
        //=======================================================================================================
        // debugging -- testpoint
        //
        private void tp(string msg) {
        }
        //
        //=======================================================================================================
        // IDisposable support
        //
        #region  IDisposable Support 
        // Do not change or add Overridable to these methods.
        // Put cleanup code in Dispose(ByVal disposing As Boolean).
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~CPDocClass() {
            Dispose(false);
            //todo  NOTE: The base class Finalize method is automatically called from the destructor:
            //base.Finalize();
        }
        #endregion
    }
}